api/v1_queues.inc

```diff
@@ -0,0 +1,161 @@
+<?php
+include_once($relPath.'user_is.inc');
+include_once("exceptions.inc");
+
+// TODO(bpfoley): There is a lot of duplicated logic betweeen this file and
+//       stats/release_queue.php's _show_round_queues
+// TODO(bpfoley): Add openAPI schema
+// NB because of how the queue_defns table is read in order, and queues
+//    that identify groups are special rows with startswith(name, '***'),
+//    we can't do pagination or sorting as part of the SQL query. The queue
+//    lists are probably short enough that we don't care. But really we should
+//    probably change the DB to have a separate queue_groups table, or some such.
+
+$user_can_see_queue_settings = $ordinary_users_can_see_queue_settings ||
+    user_is_a_sitemanager() || user_is_proj_facilitator();
+
+// DP API v1 -- Queues
+
+//---------------------------------------------------------------------------
+// queues/:roundid -- list queues in round
+function api_v1_queues($method, $data, $query_params)
+{
+    global $user_can_see_queue_settings;
+    $round = $data[":roundid"];
+    $show = get_enumerated_param($query_params, "show", "all", ["enabled", "populated", "all"]);
+
+    // Fields we  can query on
+    $valid_fields = [
+        "name",
+        "group",
+    ];
+
+    // pull out the query parameters
+    $query = [];
+    foreach (array_intersect($valid_fields, array_keys($query_params)) as $field) {
+        $param = $query_params[$field];
+        $query[$field] = is_array($param) ? $param : [$param];
+    }
+    // Building the WHERE clause is rather ad-hoc, mostly because
+    // a bunch of the output fields don't come directly from the DB result.
+    $where = sprintf("round_id='%s'", DPDatabase::escape($round->id));
+    if (in_array($show, ["enabled", "populated"])) {
+        $where .= " AND enabled = 1";
+    }
+    // The populated & group queries are filtered in the rendering loop
+    if (array_key_exists("name", $query)) {
+        $values = array_map("DPDatabase::escape", $query["name"]);
+        $likes_str = surround_and_join($values, "name LIKE '%", "%'", ' OR ');
+        $where .= " AND ($likes_str)";
+    }
+    // We always want to see the rows that introduce groups.
+    $where .= " OR (SUBSTR(name, 1, 1) = '*')";
+    $q_sql = "
+        SELECT *
+        FROM queue_defns
+        WHERE $where
+        ORDER BY ordering
+    ";
+    $result = DPDatabase::query($q_sql);
+    if (!$result) {
+        throw new ServerError(DPDatabase::log_error());
```

Comment:
We intentionally do not display DB messages to the end-user (ask me in Slack if you care to know the details on why). The `DPDatabase::query()` call above will throw an exception on query error (see https://github.com/DistributedProofreaders/dproofreaders/blob/master/pinc/DPDatabase.inc#L24 and https://github.com/DistributedProofreaders/dproofreaders/blob/master/pinc/DPDatabase.inc#L90) unless you tell it otherwise so this will never get used. In general we should rely on `DPDatabase::query()` doing The Right Thing on failure and if it returns then its return value is valid.

Reply:
Ah, interesting. I copied the pattern from `v1_projects.inc` without really understanding what it was doing.

Reply:
I am, sadly, not at all surprised that you pulled it from elsewhere in the codebase. And without `git blame` I can tell you that it was my code you copied. 😕